File: CPP/biweekly-contest-39/No2.cc
/**
 * Created by Xiaozhong on 2020/11/14.
 * Copyright (c) 2020/11/14 Xiaozhong. All rights reserved.
 */
#include <string>
#include <vector>
#include <iostream>

using namespace std;

class Solution {
public:
    int minimumDeletions(string s) {
        int n = s.size();
        vector<vector<int>> dp(2, vector<int>(n + 1));
        for (int i = 1; i <= n; ++i) {
            if (s[i - 1] == 'a') {
                dp[0][i] = dp[0][i - 1];
                dp[1][i] = dp[1][i - 1] + 1;
            } else {
                dp[0][i] = dp[0][i - 1] + 1;
                dp[1][i] = min(dp[0][i - 1], dp[1][i - 1]);
            }
        }
        return min(dp[1][n], dp[0][n]);
    }
};

int main() {
    Solution s;
    cout << s.minimumDeletions("aababbab") << endl;
}

File: Java/src/biweeklyContest57/No2.java
/*
 * Copyright (c) 2021, HuXiaozhong. All rights reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 */

package biweeklyContest57;

import java.util.*;

public class No2 {
    public int smallestChair(int[][] times, int targetFriend) {
        // 没占用的
        PriorityQueue<Integer> queue = new PriorityQueue<>();
        // 占用的
        Set<Integer> set = new HashSet<>();
        // 映射谁坐了哪个座位
        Map<Integer, Integer> mapper = new HashMap<>();
        int n = times.length;
        for (int num = 0; num < n; ++num) queue.offer(num);
        int[][] arrives = new int[n][2];
        int[][] leaves = new int[n][2];
        int start = Integer.MAX_VALUE, end = Integer.MIN_VALUE;
        for (int i = 0; i < n; ++i) {
            arrives[i][0] = times[i][0];
            arrives[i][1] = i;
            start = Math.min(start, times[i][0]);
            leaves[i][0] = times[i][1];
            leaves[i][1] = i;
            end = Math.max(end, times[i][1]);
        }
        Arrays.sort(arrives, Comparator.comparingInt(o -> o[0]));
        Arrays.sort(leaves, Comparator.comparingInt(o -> o[0]));
        int arrivesIdx = 0, leaveIdx = 0;

        for (int time = start; time <= end; ++time) {
            if (leaves[leaveIdx][0] <= time) {
                while (leaveIdx < n && leaves[leaveIdx][0] <= time) {
                    int id = leaves[leaveIdx][1];
                    int pos = mapper.get(id);
                    queue.offer(pos);
                    set.remove(pos);
                    mapper.remove(id);
                    ++leaveIdx;
                }
            }
            if (arrives[arrivesIdx][0] <= time) {
                while (arrivesIdx < n && arrives[arrivesIdx][0] <= time) {
                    int pos = queue.poll();
                    while (set.contains(pos)) pos = queue.poll();
                    if (arrives[arrivesIdx][1] == targetFriend) return pos;
                    mapper.put(arrives[arrivesIdx][1], pos);
                    set.add(pos);
                    ++arrivesIdx;
                }
            }
        }
        return -1;
    }
}
